File: src/dotnet/Ansight.Native/DefaultMemoryChannels.cs
namespace Ansight;

using System;

/// <summary>
/// Flags describing which of Ansight's built-in memory channels should be created.
/// </summary>
[Flags]
public enum DefaultMemoryChannels : byte
{
    None = 0,
    ManagedHeap = 1 << 0,
    NativeHeap = 1 << 1,
    ResidentSetSize = 1 << 2,
    PhysicalFootprint = 1 << 3,
    All = ManagedHeap | NativeHeap | ResidentSetSize | PhysicalFootprint,
#if ANDROID
    PlatformDefaults = ManagedHeap | NativeHeap | ResidentSetSize,
#elif IOS || MACCATALYST
    PlatformDefaults = ManagedHeap | PhysicalFootprint,
#else
    PlatformDefaults = ManagedHeap,
#endif
}

File: src/ios/Sources/AnsightCore/DefaultMemoryChannels.swift
import Foundation

public struct DefaultMemoryChannels: OptionSet, Sendable, Codable, Hashable {
    public let rawValue: Int

    public static let none = DefaultMemoryChannels([])
    public static let managedHeap = DefaultMemoryChannels(rawValue: 1 << 0)
    public static let nativeHeap = DefaultMemoryChannels(rawValue: 1 << 1)
    public static let residentSetSize = DefaultMemoryChannels(rawValue: 1 << 2)
    public static let physicalFootprint = DefaultMemoryChannels(rawValue: 1 << 3)
    public static let all: DefaultMemoryChannels = [.managedHeap, .nativeHeap, .residentSetSize, .physicalFootprint]
    public static let platformDefaults: DefaultMemoryChannels = [.managedHeap, .physicalFootprint]

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }
}
